File: tools/file/file.go
package file

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	oasis "github.com/nevindra/oasis"
)

// Tool provides file operations within a sandboxed workspace.
type Tool struct {
	workspacePath string
}

// New creates a FileTool restricted to workspacePath.
func New(workspacePath string) *Tool {
	return &Tool{workspacePath: workspacePath}
}

func (t *Tool) Definitions() []oasis.ToolDefinition {
	return []oasis.ToolDefinition{
		{
			Name:        "file_read",
			Description: "Read a file from the workspace. Returns the file content (truncated to 8000 chars if large).",
			Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"File path relative to workspace"}},"required":["path"]}`),
		},
		{
			Name:        "file_write",
			Description: "Write content to a file in the workspace. Creates parent directories if needed.",
			Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"File path relative to workspace"},"content":{"type":"string","description":"Content to write"}},"required":["path","content"]}`),
		},
		{
			Name:        "file_list",
			Description: "List files and directories in a workspace directory. Returns one entry per line with type prefix (file/dir) and name.",
			Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"Directory path relative to workspace (empty or '.' for root)"}}}`),
		},
		{
			Name:        "file_delete",
			Description: "Delete a file or empty directory from the workspace.",
			Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"File or directory path relative to workspace"}},"required":["path"]}`),
		},
		{
			Name:        "file_stat",
			Description: "Get metadata for a file or directory in the workspace. Returns name, size, type, and modification time.",
			Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"File or directory path relative to workspace"}},"required":["path"]}`),
		},
	}
}

func (t *Tool) Execute(ctx context.Context, name string, args json.RawMessage) (oasis.ToolResult, error) {
	var params struct {
		Path    string `json:"path"`
		Content string `json:"content"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return oasis.ToolResult{Error: "invalid args: " + err.Error()}, nil
	}

	path := params.Path
	if path == "" {
		path = "."
	}
	resolved, err := t.resolvePath(path)
	if err != nil {
		return oasis.ToolResult{Error: err.Error()}, nil
	}

	switch name {
	case "file_read":
		return t.read(resolved)
	case "file_write":
		return t.write(resolved, params.Content)
	case "file_list":
		return t.list(resolved)
	case "file_delete":
		return t.remove(resolved)
	case "file_stat":
		return t.stat(resolved)
	default:
		return oasis.ToolResult{Error: "unknown file tool: " + name}, nil
	}
}

func (t *Tool) resolvePath(path string) (string, error) {
	if filepath.IsAbs(path) {
		return "", fmt.Errorf("absolute paths not allowed: %s", path)
	}
	if strings.Contains(path, "..") {
		return "", fmt.Errorf("path traversal not allowed: %s", path)
	}
	resolved := filepath.Join(t.workspacePath, path)
	// Double-check it's still within workspace
	if !strings.HasPrefix(resolved, t.workspacePath) {
		return "", fmt.Errorf("path escapes workspace: %s", path)
	}
	return resolved, nil
}

func (t *Tool) read(path string) (oasis.ToolResult, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return oasis.ToolResult{Error: "read error: " + err.Error()}, nil
	}
	content := string(data)
	if len(content) > 8000 {
		content = content[:8000] + "\n... (truncated)"
	}
	return oasis.ToolResult{Content: content}, nil
}

func (t *Tool) write(path, content string) (oasis.ToolResult, error) {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return oasis.ToolResult{Error: "mkdir error: " + err.Error()}, nil
	}
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		return oasis.ToolResult{Error: "write error: " + err.Error()}, nil
	}
	return oasis.ToolResult{Content: fmt.Sprintf("Written %d bytes to %s", len(content), filepath.Base(path))}, nil
}

func (t *Tool) list(path string) (oasis.ToolResult, error) {
	entries, err := os.ReadDir(path)
	if err != nil {
		return oasis.ToolResult{Error: "list error: " + err.Error()}, nil
	}
	var b strings.Builder
	for _, e := range entries {
		kind := "file"
		if e.IsDir() {
			kind = "dir"
		}
		fmt.Fprintf(&b, "%s\t%s\n", kind, e.Name())
	}
	return oasis.ToolResult{Content: b.String()}, nil
}

func (t *Tool) remove(path string) (oasis.ToolResult, error) {
	if err := os.Remove(path); err != nil {
		return oasis.ToolResult{Error: "delete error: " + err.Error()}, nil
	}
	return oasis.ToolResult{Content: fmt.Sprintf("Deleted %s", filepath.Base(path))}, nil
}

func (t *Tool) stat(path string) (oasis.ToolResult, error) {
	info, err := os.Stat(path)
	if err != nil {
		return oasis.ToolResult{Error: "stat error: " + err.Error()}, nil
	}
	kind := "file"
	if info.IsDir() {
		kind = "directory"
	}
	out, _ := json.Marshal(map[string]any{
		"name":     info.Name(),
		"size":     info.Size(),
		"type":     kind,
		"modified": info.ModTime().UTC().Format("2006-01-02T15:04:05Z"),
	})
	return oasis.ToolResult{Content: string(out)}, nil
}

File: crates/oasis-brain/src/tool/file.rs
use async_trait::async_trait;
use oasis_core::types::ToolDefinition;
use serde_json::json;

use crate::tool::{Tool, ToolResult};

const MAX_WRITE_SIZE: usize = 1_048_576; // 1MB
const MAX_READ_SIZE: usize = 1_048_576; // 1MB

/// File extensions considered sensitive — blocked even inside workspace.
const SENSITIVE_EXTENSIONS: &[&str] = &[".env", ".key", ".pem", ".p12", ".pfx", ".jks"];

pub struct FileTool {
    workspace: String,
}

impl FileTool {
    pub fn new(workspace: String) -> Self {
        Self { workspace }
    }

    /// Resolve a path to an absolute path within the workspace.
    /// Returns Err if the path escapes the workspace root.
    fn resolve_path(&self, path: &str) -> Result<std::path::PathBuf, String> {
        let workspace = std::path::Path::new(&self.workspace);

        // Workspace must exist for safe path resolution
        let canonical_workspace = workspace
            .canonicalize()
            .map_err(|_| "Workspace directory does not exist".to_string())?;

        // Reject absolute paths that don't start with the workspace
        if std::path::Path::new(path).is_absolute() {
            let abs = std::path::PathBuf::from(path);
            if !abs.starts_with(&canonical_workspace) {
                return Err("Access denied: path escapes workspace root".to_string());
            }
        }

        let resolved = canonical_workspace.join(path);

        // For existing paths, canonicalize and check containment.
        // For new paths, canonicalize the parent and check containment.
        let canonical = if resolved.exists() {
            resolved.canonicalize().map_err(|e| format!("Invalid path: {e}"))?
        } else {
            // Canonicalize the parent, then append the filename
            let parent = resolved.parent().unwrap_or(&canonical_workspace);
            if !parent.exists() {
                // Try to resolve what we can — walk components from workspace root
                let mut check = canonical_workspace.clone();
                let rel = resolved.strip_prefix(&canonical_workspace).unwrap_or(&resolved);
                for component in rel.components() {
                    check.push(component);
                    if check.exists() {
                        check = check.canonicalize().map_err(|e| format!("Invalid path: {e}"))?;
                    }
                }
                check
            } else {
                let canonical_parent = parent.canonicalize().map_err(|e| format!("Invalid path: {e}"))?;
                let filename = resolved.file_name().ok_or("Invalid filename")?;
                canonical_parent.join(filename)
            }
        };

        if !canonical.starts_with(&canonical_workspace) {
            return Err("Access denied: path escapes workspace root".to_string());
        }

        // Check sensitive file extensions
        if let Some(name) = canonical.file_name().and_then(|n| n.to_str()) {
            let lower = name.to_lowercase();
            for ext in SENSITIVE_EXTENSIONS {
                if lower.ends_with(ext) {
                    return Err(format!("Access denied: sensitive file type ({ext})"));
                }
            }
        }

        Ok(canonical)
    }
}

#[async_trait]
impl Tool for FileTool {
    fn definitions(&self) -> Vec<ToolDefinition> {
        vec![
            ToolDefinition {
                name: "file_read".to_string(),
                description: "Read the contents of a file in the workspace.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "path": { "type": "string", "description": "File path (relative to workspace root)" }
                    },
                    "required": ["path"]
                }),
            },
            ToolDefinition {
                name: "file_write".to_string(),
                description: "Write content to a file in the workspace. Creates parent directories if needed.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "path": { "type": "string", "description": "File path (relative to workspace root)" },
                        "content": { "type": "string", "description": "Content to write" }
                    },
                    "required": ["path", "content"]
                }),
            },
            ToolDefinition {
                name: "file_list".to_string(),
                description: "List files and directories in a workspace path.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "path": { "type": "string", "description": "Directory path (relative to workspace root). Defaults to workspace root." }
                    }
                }),
            },
        ]
    }

    async fn execute(&self, name: &str, args: &serde_json::Value) -> ToolResult {
        match name {
            "file_read" => {
                let path = match args["path"].as_str() {
                    Some(p) => p,
                    None => return ToolResult::err("Missing required parameter: path"),
                };
                let resolved = match self.resolve_path(path) {
                    Ok(p) => p,
                    Err(e) => return ToolResult::err(e),
                };
                match tokio::fs::read_to_string(&resolved).await {
                    Ok(content) => {
                        if content.len() > MAX_READ_SIZE {
                            ToolResult::ok(format!(
                                "{}...\n(truncated, {} bytes total)",
                                &content[..MAX_READ_SIZE],
                                content.len()
                            ))
                        } else {
                            ToolResult::ok(content)
                        }
                    }
                    Err(e) => ToolResult::err(format!("Failed to read file: {e}")),
                }
            }
            "file_write" => {
                let path = match args["path"].as_str() {
                    Some(p) => p,
                    None => return ToolResult::err("Missing required parameter: path"),
                };
                let content = match args["content"].as_str() {
                    Some(c) => c,
                    None => return ToolResult::err("Missing required parameter: content"),
                };
                if content.len() > MAX_WRITE_SIZE {
                    return ToolResult::err(format!(
                        "Content too large: {} bytes (max {})",
                        content.len(),
                        MAX_WRITE_SIZE
                    ));
                }
                let resolved = match self.resolve_path(path) {
                    Ok(p) => p,
                    Err(e) => return ToolResult::err(e),
                };
                // Create parent directories
                if let Some(parent) = resolved.parent() {
                    let _ = tokio::fs::create_dir_all(parent).await;
                }
                match tokio::fs::write(&resolved, content).await {
                    Ok(()) => ToolResult::ok(format!(
                        "Written {} bytes to {}",
                        content.len(),
                        resolved.display()
                    )),
                    Err(e) => ToolResult::err(format!("Failed to write file: {e}")),
                }
            }
            "file_list" => {
                let path = args["path"].as_str().unwrap_or(".");
                let resolved = match self.resolve_path(path) {
                    Ok(p) => p,
                    Err(e) => return ToolResult::err(e),
                };
                let mut entries = match tokio::fs::read_dir(&resolved).await {
                    Ok(rd) => rd,
                    Err(e) => return ToolResult::err(format!("Failed to list directory: {e}")),
                };
                let mut output = Vec::new();
                while let Ok(Some(entry)) = entries.next_entry().await {
                    let name = entry.file_name().to_string_lossy().to_string();
                    let is_dir = entry.file_type().await.map(|t| t.is_dir()).unwrap_or(false);
                    if is_dir {
                        output.push(format!("{name}/"));
                    } else {
                        output.push(name);
                    }
                }
                output.sort();
                if output.is_empty() {
                    ToolResult::ok("(empty directory)")
                } else {
                    ToolResult::ok(output.join("\n"))
                }
            }
            _ => ToolResult::err(format!("Unknown file tool: {name}")),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn setup_workspace() -> std::path::PathBuf {
        let dir = std::env::temp_dir().join("oasis-file-tool-test");
        let _ = std::fs::create_dir_all(&dir);
        dir
    }

    #[test]
    fn test_path_traversal_blocked() {
        let ws = setup_workspace();
        let tool = FileTool::new(ws.to_string_lossy().to_string());
        assert!(tool.resolve_path("../../etc/passwd").is_err());
        assert!(tool.resolve_path("/etc/passwd").is_err());
    }

    #[test]
    fn test_sensitive_files_blocked() {
        let ws = setup_workspace();
        let tool = FileTool::new(ws.to_string_lossy().to_string());
        assert!(tool.resolve_path("config.env").is_err());
        assert!(tool.resolve_path("server.key").is_err());
        assert!(tool.resolve_path("cert.pem").is_err());
    }

    #[test]
    fn test_valid_paths_allowed() {
        let ws = setup_workspace();
        let tool = FileTool::new(ws.to_string_lossy().to_string());
        let result = tool.resolve_path("myapp/src/main.rs");
        // Either Ok (path resolves within workspace) or non-security error
        if let Err(e) = result {
            assert!(!e.contains("Access denied"), "Should not be a security error: {e}");
        }
    }

    #[test]
    fn test_nonexistent_workspace_rejected() {
        let tool = FileTool::new("/nonexistent/workspace/path".to_string());
        assert!(tool.resolve_path("anything.txt").is_err());
    }
}
